Hash fields used in task service IDs

schmichael opened the pull request:
Fixes #3620

Previously we concatenated tags into task service IDs. This could break
deregistration of tag names that contained double //s like some Fabio
tags.

This change breaks service ID backward compatibility so on upgrade all
users services and checks will be removed and re-added with new IDs.

This change has the side effect of including all service fields in the
ID's hash, so we no longer have to track PortLabel and AddressMode
changes independently.

jippi commented:
What format have they changed from and to?

dadgar commented on command/agent/consul/client.go:
```diff
@@ -999,36 +1003,40 @@ func (c *ServiceClient) removeTaskRegistration(allocID, taskName string) {
-	parts[4] = service.Name
-	copy(parts[5:], service.Tags)
-	return strings.Join(parts, "-")
+	h := sha1.New()
```
Hmm why not keep the hash as a method on Service and have `return nomadTaskPrefix + service.Hash()`

preetapan replied:
+1 on this. No reason to maintain all the old fields in the hash, all we care about is that this creates a unique id for the service.

schmichael commented:
> What format have they changed from and to? 
> -- @jippi 

The format is intentionally not documented outside of code because it is not something users should have to be aware of or rely on. That being said the old and new formats are here in the code:

https://github.com/hashicorp/nomad/pull/3632/files#diff-8612d5e61f84976db572b3f38eb380c0L1015

schmichael commented on command/agent/consul/client.go:
```diff
@@ -389,38 +401,31 @@ func (c *ServiceClient) sync() error {
 		// Unknown Nomad managed service; kill
 		if err := c.client.ServiceDeregister(id); err != nil {
+			if isOldNomadService(id) {
+				// Don't hard-fail on old entries. See #3620
```
Yeah, I think the optimal solution is soft-failing on all deregistration calls, logging only the first failure per-id, and then logging if it ever succeeds... I guess that would only take a map of IDs to track... should I add that and drop `isOldNomadService` entirely?

dadgar replied:
Yeah I wasn't asking for an action on this one. I think it is just unfortunate we will have to maintain this behavior for a while!

schmichael commented:
Addressed issues and am merging into the parent branch `f-3380-custom-ports`, so I don't shoot myself in the foot by rebasing the intermediary branch (again)
